### PR TITLE
refactor: Refactor lifecyclehandler test to avoid circular dependency

### PR DESCRIPTION
**Pull request by justinsb**
This test was using the full controller test machinery, which caused
circular dependencies when we used the lifecycle handler in
controllers.

### CODE REVIEW

File: pkg/test/kubeharness.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/mockkubeapiserver"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/dynamic"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/crd/crdloader"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/logging"
+)
+
+// KubeHarness is a test harness that brings up a kube-apiserver (only).
```
Comment (justinsb) [inline pkg/test/kubeharness.go:45]: This is a subset of our full harness.  I hope we can share the code in future, but wanted to keep this PR short.

File: pkg/test/kubeharness.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		}
+
+		t.Cleanup(func() {
+			if err := env.Stop(); err != nil {
+				h.Errorf("error stopping envtest environment: %v", err)
+			}
+		})
+
+		h.restConfig = restConfig
+	} else if targetKube := os.Getenv("E2E_KUBE_TARGET"); targetKube == "mock" {
```
Comment (yuwenma) [inline pkg/test/kubeharness.go:89]: Shall this be `E2E_GCP_TARGET`?

File: pkg/test/kubeharness.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+func (h *KubeHarness) waitForCRDReady(obj client.Object) {
+	logger := log.FromContext(h.Ctx)
+
+	apiVersion, kind := obj.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+
+	id := types.NamespacedName{Name: name, Namespace: namespace}
+	if err := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
```
Comment (yuwenma) [inline pkg/test/kubeharness.go:196]: The `PollImmediate` is deprecated. The lib suggests to use `PollWithContextTimeout` instead.

File: pkg/test/kubeharness.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	ns.SetName(name)
+	if err := h.GetClient().Create(ctx, ns); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			h.Fatalf("error creating namespace %v: %v", name, err)
+		}
+	}
+}
+
+// CreateDummyCRD registers a CRD so we can create objects in tests
+func (h *KubeHarness) CreateDummyCRD(group, version, kind string) {
```
Comment (yuwenma) [inline pkg/test/kubeharness.go:238]: This is cool. (What do we do today?  I guess we should have such a CreateCRD already, but cannot find it via code search)

**Comment by yuwenma** [comment]
/lgtm
/approve

I like the direction of this PR.

File: pkg/test/kubeharness.go (hunk 1 above)
Comment (yuwenma) [reply pkg/test/kubeharness.go:45]: Yes! And +1 for smaller PRs

File: pkg/test/kubeharness.go (hunk 2 above)
Comment (justinsb) [reply pkg/test/kubeharness.go:89]: I don't believe so, this is to enable (fast) mock-kubeapiserver instead of (slow) envtest

Comment (yuwenma) [reply pkg/test/kubeharness.go:89]: ohoh, I see. Both envs use value "mock". 🤣

File: pkg/test/kubeharness.go (hunk 4 above)
Comment (justinsb) [reply pkg/test/kubeharness.go:238]: I found it ... there's a function called `test.FakeCRDs()` that is called via a convoluted call chain.

I'll harmonize this implement in a follow-on PR.

**Pull request merged**